### PR TITLE
Rholang: Extract out wildcards from freeCount

### PR DESCRIPTION
**Pull request by KentShikama**
wildcard is now a boolean returning method in
HasLocallyFree.

This closes https://rchain.atlassian.net/browse/RHOL-188

### CODE REVIEW

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 1]
```diff
@@ -200,16 +200,51 @@ object implicits {
         that.freeCount + p.freeCount,
         that.locallyFree | p.locallyFree
       )
+
+    // TODO: Receives and matches
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Please remove this TODO

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 2]
```diff
@@ -200,16 +200,51 @@ object implicits {
...
+    def wildcard: Boolean =
+      p.sends.map(send => SendLocallyFree.wildcard(send)).exists(identity) || p.evals
+        .map(eval => EvalLocallyFree.wildcard(eval))
+        .exists(identity) || p.exprs.map(expr => ExprLocallyFree.wildcard(expr)).exists(identity)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: The fact the the best way is to use exists(identity) is a giant red mark against scala.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala (hunk 1 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala (hunk 2 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Hmm what is the Haskell way to do this?

Also I'm not sure if this is the way its supposed to be in Scala. Expert opinions @henrytill ?

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: To give some context, I'm folding over a list of booleans using OR .

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 3]
```diff
@@ -200,16 +200,86 @@ object implicits {
         that.freeCount + p.freeCount,
         that.locallyFree | p.locallyFree
       )
+
+    def wildcard: Boolean =
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: No, for Par, just put it in the proto.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 4]
```diff
@@ -200,16 +200,86 @@ object implicits {
...
+        .exists(identity) || p.matches
+        .map(_match => MatchLocallyFree.wildcard(_match))
+        .exists(identity)
   }
 
   implicit def fromPar[T](p: T)(implicit toPar: T => Par): Option[Par] = Some(p)
 
   implicit val SendLocallyFree: HasLocallyFree[Send] = new HasLocallyFree[Send] {
+    def wildcard(s: Send) =
+      ChannelLocallyFree.wildcard(s.chan.get) || s.data.map(_.wildcard).exists(identity)
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Same here. Any of the protos that have freeCount explicitly should have wildcard there too.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 5]
```diff
@@ -200,16 +200,86 @@ object implicits {
...
+      val wildcardBinds = r.binds.foldLeft(false) { (acc, bind: ReceiveBind) =>
+        bind match {
+          case ReceiveBind(patterns: Seq[Channel], source: Option[Channel]) =>
+            acc || patterns
+              .map(pattern => ChannelLocallyFree.wildcard(pattern))
+              .exists(identity) || ChannelLocallyFree.wildcard(source.get)
+        }
+      }
+      r.body.get.wildcard || wildcardBinds
+    }
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Same here.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala [hunk 6]
```diff
@@ -200,16 +200,86 @@ object implicits {
...
+  implicit val MatchLocallyFree: HasLocallyFree[Match] = new HasLocallyFree[Match] {
+    def wildcard(m: Match) = {
+      val wildcardMatchCases = m.cases.foldLeft(false) { (acc, _case: MatchCase) =>
+        _case match {
+          case MatchCase(pattern: Option[Par], source: Option[Par]) =>
+            acc || pattern.get.wildcard || source.get.wildcard
+        }
+      }
+      m.target.get.wildcard || wildcardMatchCases
+    }
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Same here.

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala (hunk 5 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala (hunk 4 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala (hunk 3 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/implicits.scala]: Done

**Comment by KentShikama** [comment]
Ready for another review.

**Pull request merged**